packages/cdktf-cli/lib/cdktf-stack.ts

```diff
@@ -266,83 +132,201 @@ export class CdktfStack {
-      return undefined;
-    }
 
-    return this.waitOnMachineDone();
+  private notifyState(
+    update:
+      | StackUpdate
+      | StackApprovalUpdate
+      | { type: "idle" }
+      | { type: "done" }
+      | { type: "error" }
+  ) {
+    logger.debug(`[${this.stackName}]: ${update.type}`);
+    this.currentState = update.type;
+    switch (update.type) {
+      case "idle":
+      case "done":
+      case "error":
+        break;
+
+      case "outputs fetched":
+      case "deployed":
+        logger.debug(`Outputs: ${JSON.stringify(update.outputs)}`);
+        logger.debug(
+          `OutputsByConstructId: ${JSON.stringify(update.outputsByConstructId)}`
+        );
+        this.outputs = update.outputs;
+        this.outputsByConstructId = update.outputsByConstructId;
+        this.context.onUpdate(update);
+        break;
+
+      default:
+        this.context.onUpdate(update);
+        break;
+    }
   }
 
-  private waitOnMachineDone(): Promise<void> {
-    return new Promise((resolve, reject) => {
-      if (this.currentState === "done") {
-        resolve();
-      }
-
-      if (this.currentState === "error") {
-        reject(new Error(this.stateMachine.state.context.message));
+  private logCallback(
+    stateName: string
+  ): (message: string, isError?: boolean) => void {
+    const onLog = this.context.onLog;
+    return (msg: string, isError = false) => {
+      const message = extractJsonLogIfPresent(msg);
+      logger.debug(`[${this.context.stack.name}](${stateName}): ${msg}`);
+      if (onLog) {
+        onLog({ message, isError });
       }
+    };
+  }
 
-      this.stateMachine.onTransition((state) => {
-        if (state.matches("error")) {
-          reject(new Error(state.context.message));
-        } else if (state.matches("done")) {
-          resolve();
-        }
+  private waitForApproval(plan: TerraformPlan) {
+    return new Promise<boolean>((resolve) => {
+      this.notifyState({
+        type: "waiting for stack approval",
+        stackName: this.stack.name,
+        plan: plan,
+        approve: () => {
+          resolve(true);
+        },
+        reject: () => {
+          resolve(false);
+        },
       });
     });
   }
 
-  public async diff(stackName?: string) {
-    this.stateMachine.send({
-      type: "START",
-      targetAction: "diff",
-      targetStack: stackName,
-    });
+  private async initalizeTerraform({
+    isSpeculative,
+  }: {
+    isSpeculative: boolean;
+  }) {
+    const terraform = await getTerraformClient(
+      this.context.abortSignal,
+      this.context.stack,
+      isSpeculative,
+      this.logCallback.bind(this)
+    );
+
+    await terraform.init();
 
-    return this.waitOnMachineDone();
+    return terraform;
   }
 
-  public async deploy(stackName?: string) {
-    this.stateMachine.send({
-      type: "START",
-      targetAction: "deploy",
-      targetStack: stackName,
-    });
+  private async handleState(cb: () => Promise<void>) {
```

Comment:
I'd propose something like `handleExecution()` or even a short `run()` because it also handles the `currentWorkPromise` and checks `this.stopped`. While it does handle state, it only handles the "run" state part of all the types that `currentState` may hold. `handleRunState` would also differentiate this enough for me.